File: days/day_06/solutions/solution.ex
defmodule Solution do
  def solve(input_data) do
    lines = String.split(String.trim(input_data), "\n")
    if lines == [], do: {0, 0}, else: do_solve(lines)
  end
  
  defp do_solve(lines) do
    # Find maximum line length and pad all lines
    max_len = lines |> Enum.map(&String.length/1) |> Enum.max()
    padded_lines = lines |> Enum.map(&String.pad_trailing(&1, max_len))
    
    # Operator row is the last row
    op_row_idx = length(padded_lines) - 1
    op_row = Enum.at(padded_lines, op_row_idx)
    num_rows = Enum.take(padded_lines, op_row_idx)
    
    # Find problem boundaries (columns that are all spaces)
    is_space_col = 0..(max_len - 1)
      |> Enum.map(fn col ->
        Enum.all?(padded_lines, fn line ->
          col >= String.length(line) || String.at(line, col) == " "
        end)
      end)
    
    # Group columns into problems
    problems = find_problems(0, max_len, is_space_col, op_row, [])
    
    # Part 1: Extract numbers horizontally
    part1_total = problems
      |> Enum.map(fn {start_col, end_col, op} ->
        numbers = num_rows
          |> Enum.flat_map(fn row ->
            problem_str = row
              |> String.slice(start_col, end_col - start_col)
              |> String.trim()
            problem_str
              |> String.split(~r/\s+/)
              |> Enum.filter(&(&1 != ""))
              |> Enum.map(&String.to_integer/1)
          end)
        
        if numbers != [] do
          result = case op do
            "+" -> Enum.sum(numbers)
            "*" -> Enum.product(numbers)
          end
          result
        else
          0
        end
      end)
      |> Enum.sum()
    
    # Part 2: Parse vertically (columns, right-to-left)
    # Approach: Use Elixir pipes and Enum functions
    part2_total = problems
      |> Enum.map(fn {start_col, end_col, op} ->
        # Extract column strings using pipes
        col_strings = start_col..(end_col - 1)
          |> Enum.reject(fn col -> Enum.at(is_space_col, col) end)
          |> Enum.map(fn col ->
            num_rows
              |> Enum.map(fn row -> String.at(row, col) || " " end)
              |> Enum.filter(fn ch -> Regex.match?(~r/[\d\s]/, ch) end)
              |> Enum.join("")
              |> String.trim()
          end)
          |> Enum.filter(&(&1 != ""))
          |> Enum.reverse()  # Right-to-left reading
        
        # Parse numbers
        numbers = col_strings
          |> Enum.map(&String.replace(&1, ~r/\s/, ""))
          |> Enum.filter(&(&1 != ""))
          |> Enum.map(&String.to_integer/1)
        
        # Apply operator
        if numbers != [] do
          result = case op do
            "+" -> Enum.sum(numbers)
            "*" -> Enum.product(numbers)
          end
          result
        else
          0
        end
      end)
      |> Enum.sum()
    
    {part1_total, part2_total}
  end
  
  defp find_problems(i, max_len, _is_space_col, _op_row, acc) when i >= max_len, do: acc
  
  defp find_problems(i, max_len, is_space_col, op_row, acc) do
    if Enum.at(is_space_col, i) do
      find_problems(i + 1, max_len, is_space_col, op_row, acc)
    else
      start_col = i
      {end_col, new_i} = find_problem_end(i, max_len, is_space_col)
      
      op = op_row
        |> String.slice(start_col, end_col - start_col)
        |> String.to_charlist()
        |> Enum.find(fn ch -> ch == ?+ || ch == ?* end)
      
      new_acc = if op do
        [{start_col, end_col, List.to_string([op])} | acc]
      else
        acc
      end
      
      find_problems(new_i, max_len, is_space_col, op_row, new_acc) |> Enum.reverse()
    end
  end
  
  defp find_problem_end(i, max_len, is_space_col) do
    if i >= max_len || Enum.at(is_space_col, i) do
      {i, i}
    else
      find_problem_end(i + 1, max_len, is_space_col)
    end
  end
end

content = File.read!("../data/input.txt")
{part1, part2} = Solution.solve(content)
IO.puts("Part 1: #{part1}")
IO.puts("Part 2: #{part2}")

File: days/day_06/solutions/solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

class Solution {
    static class Problem {
        int startCol;
        int endCol;
        char op;
        
        Problem(int startCol, int endCol, char op) {
            this.startCol = startCol;
            this.endCol = endCol;
            this.op = op;
        }
    }
    
    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get("../data/input.txt");
        String content = new String(Files.readAllBytes(inputPath));
        long[] results = solve(content);
        System.out.println("Part 1: " + results[0]);
        System.out.println("Part 2: " + results[1]);
    }
    
    private static long[] solve(String inputData) {
        String[] lines = inputData.trim().split("\n");
        if (lines.length == 0) {
            return new long[]{0, 0};
        }
        
        // Find maximum line length and pad all lines
        int maxLen = Arrays.stream(lines).mapToInt(String::length).max().orElse(0);
        String[] paddedLines = Arrays.stream(lines)
            .map(line -> String.format("%-" + maxLen + "s", line))
            .toArray(String[]::new);
        
        // Operator row is the last row
        int opRowIdx = paddedLines.length - 1;
        String opRow = paddedLines[opRowIdx];
        String[] numRows = Arrays.copyOf(paddedLines, opRowIdx);
        
        // Part 1: Parse horizontally
        long part1Total = 0;
        
        // Find problem boundaries (columns that are all spaces)
        boolean[] isSpaceCol = new boolean[maxLen];
        for (int col = 0; col < maxLen; col++) {
            final int c = col;
            isSpaceCol[col] = Arrays.stream(paddedLines)
                .allMatch(line -> c >= line.length() || line.charAt(c) == ' ');
        }
        
        // Group columns into problems
        List<Problem> problems = new ArrayList<>();
        int i = 0;
        while (i < maxLen) {
            if (!isSpaceCol[i]) {
                int startCol = i;
                while (i < maxLen && !isSpaceCol[i]) {
                    i++;
                }
                int endCol = i;
                // Extract operator for this problem
                char op = 0;
                for (int j = startCol; j < endCol; j++) {
                    if (j < opRow.length() && (opRow.charAt(j) == '+' || opRow.charAt(j) == '*')) {
                        op = opRow.charAt(j);
                        break;
                    }
                }
                if (op != 0) {
                    problems.add(new Problem(startCol, endCol, op));
                }
            } else {
                i++;
            }
        }
        
        // Solve Part 1: Extract numbers horizontally
        for (Problem prob : problems) {
            List<Long> numbers = new ArrayList<>();
            for (String row : numRows) {
                if (prob.endCol > row.length()) continue;
                String problemStr = row.substring(prob.startCol, prob.endCol).trim();
                String[] parts = problemStr.split("\\s+");
                for (String part : parts) {
                    try {
                        numbers.add(Long.parseLong(part));
                    } catch (NumberFormatException e) {
                        // Skip invalid numbers
                    }
                }
            }
            
            // Apply operator
            if (!numbers.isEmpty()) {
                long result;
                if (prob.op == '+') {
                    result = numbers.stream().mapToLong(Long::longValue).sum();
                } else {
                    result = numbers.stream().reduce(1L, (a, b) -> a * b);
                }
                part1Total += result;
            }
        }
        
        // Part 2: Parse vertically (columns, right-to-left)
        // Approach: Use Java streams to transpose and extract numbers
        long part2Total = 0;
        
        for (Problem prob : problems) {
            // Extract column strings (transpose)
            List<String> colStrings = new ArrayList<>();
            for (int col = prob.startCol; col < prob.endCol; col++) {
                if (col < isSpaceCol.length && isSpaceCol[col]) continue;
                
                StringBuilder colStr = new StringBuilder();
                for (String row : numRows) {
                    if (col < row.length()) {
                        char ch = row.charAt(col);
                        if (Character.isDigit(ch) || ch == ' ') {
                            colStr.append(ch);
                        }
                    }
                }
                String trimmed = colStr.toString().trim();
                if (!trimmed.isEmpty()) {
                    colStrings.add(trimmed);
                }
            }
            
            // Parse numbers from columns (reading right-to-left means reverse)
            // Reverse for right-to-left reading, then parse
            Collections.reverse(colStrings);
            List<Long> numbers = colStrings.stream()
                .map(s -> s.replaceAll(" ", ""))
                .filter(s -> !s.isEmpty())
                .map(Long::parseLong)
                .collect(java.util.stream.Collectors.toList());
            
            // Apply operator
            if (!numbers.isEmpty()) {
                long result;
                if (prob.op == '+') {
                    result = numbers.stream().mapToLong(Long::longValue).sum();
                } else {
                    result = numbers.stream().reduce(1L, (a, b) -> a * b);
                }
                part2Total += result;
            }
        }
        
        return new long[]{part1Total, part2Total};
    }
}
